File: py_client/gnitz_client/expr.py
"""Client-side expression bytecode helper for filter predicates.

Mirrors the server-side ExprProgram 4-word encoding: (opcode, dst, arg1, arg2).
Produces dicts with 'num_regs', 'result_reg', 'code' ready for packing
into PARAM_EXPR_BASE slots on CircuitBuilder filter nodes.
"""

# Expression opcodes (must match gnitz/dbsp/expr.py)
EXPR_LOAD_COL_INT = 1
EXPR_LOAD_COL_FLOAT = 2
EXPR_LOAD_CONST = 3
EXPR_INT_ADD = 4
EXPR_INT_SUB = 5
EXPR_CMP_EQ = 15
EXPR_CMP_NE = 16
EXPR_CMP_GT = 17
EXPR_CMP_GE = 18
EXPR_CMP_LT = 19
EXPR_CMP_LE = 20
EXPR_BOOL_AND = 27
EXPR_BOOL_OR = 28
EXPR_BOOL_NOT = 29
EXPR_IS_NULL = 30
EXPR_IS_NOT_NULL = 31


class ExprBuilder:
    """Ergonomic builder for expression bytecode (client-side)."""

    def __init__(self):
        self._code: list[int] = []
        self._next_reg = 0

    def _alloc_reg(self) -> int:
        r = self._next_reg
        self._next_reg += 1
        return r

    def _emit(self, op: int, dst: int, a1: int, a2: int):
        self._code.extend([op, dst, a1, a2])

    def load_col_int(self, col_idx: int) -> int:
        dst = self._alloc_reg()
        self._emit(EXPR_LOAD_COL_INT, dst, col_idx, 0)
        return dst

    def load_const(self, value: int) -> int:
        dst = self._alloc_reg()
        self._emit(EXPR_LOAD_CONST, dst, value, 0)
        return dst

    def cmp_gt(self, a: int, b: int) -> int:
        dst = self._alloc_reg()
        self._emit(EXPR_CMP_GT, dst, a, b)
        return dst

    def cmp_eq(self, a: int, b: int) -> int:
        dst = self._alloc_reg()
        self._emit(EXPR_CMP_EQ, dst, a, b)
        return dst

    def cmp_lt(self, a: int, b: int) -> int:
        dst = self._alloc_reg()
        self._emit(EXPR_CMP_LT, dst, a, b)
        return dst

    def cmp_ge(self, a: int, b: int) -> int:
        dst = self._alloc_reg()
        self._emit(EXPR_CMP_GE, dst, a, b)
        return dst

    def cmp_le(self, a: int, b: int) -> int:
        dst = self._alloc_reg()
        self._emit(EXPR_CMP_LE, dst, a, b)
        return dst

    def cmp_ne(self, a: int, b: int) -> int:
        dst = self._alloc_reg()
        self._emit(EXPR_CMP_NE, dst, a, b)
        return dst

    def bool_and(self, a: int, b: int) -> int:
        dst = self._alloc_reg()
        self._emit(EXPR_BOOL_AND, dst, a, b)
        return dst

    def bool_or(self, a: int, b: int) -> int:
        dst = self._alloc_reg()
        self._emit(EXPR_BOOL_OR, dst, a, b)
        return dst

    def bool_not(self, a: int) -> int:
        dst = self._alloc_reg()
        self._emit(EXPR_BOOL_NOT, dst, a, 0)
        return dst

    def is_not_null(self, col_idx: int) -> int:
        dst = self._alloc_reg()
        self._emit(EXPR_IS_NOT_NULL, dst, col_idx, 0)
        return dst

    def is_null(self, col_idx: int) -> int:
        dst = self._alloc_reg()
        self._emit(EXPR_IS_NULL, dst, col_idx, 0)
        return dst

    def build(self, result_reg: int) -> dict:
        return {
            'num_regs': self._next_reg,
            'result_reg': result_reg,
            'code': list(self._code),
        }

File: rust_client/gnitz-core/src/expr.rs
// Expression opcodes — must match gnitz/dbsp/expr.py
const EXPR_LOAD_COL_INT:   u32 = 1;
const EXPR_LOAD_COL_FLOAT: u32 = 2;
const EXPR_LOAD_CONST:     u32 = 3;
const EXPR_INT_ADD:        u32 = 4;
const EXPR_INT_SUB:        u32 = 5;
const EXPR_CMP_EQ:         u32 = 15;
const EXPR_CMP_NE:         u32 = 16;
const EXPR_CMP_GT:         u32 = 17;
const EXPR_CMP_GE:         u32 = 18;
const EXPR_CMP_LT:         u32 = 19;
const EXPR_CMP_LE:         u32 = 20;
const EXPR_BOOL_AND:       u32 = 27;
const EXPR_BOOL_OR:        u32 = 28;
const EXPR_BOOL_NOT:       u32 = 29;
const EXPR_IS_NULL:        u32 = 30;
const EXPR_IS_NOT_NULL:    u32 = 31;

/// A compiled expression program: a flat list of 4-word instructions
/// (opcode, dst_reg, arg1, arg2) plus metadata for embedding in filter params.
#[derive(Clone, Debug)]
pub struct ExprProgram {
    pub num_regs:   u32,
    pub result_reg: u32,
    pub code:       Vec<u32>,
}

/// Builds an expression bytecode program with automatic register allocation.
pub struct ExprBuilder {
    code:     Vec<u32>,
    next_reg: u32,
}

impl ExprBuilder {
    pub fn new() -> Self {
        ExprBuilder { code: Vec::new(), next_reg: 0 }
    }

    fn alloc_reg(&mut self) -> u32 {
        let r = self.next_reg;
        self.next_reg += 1;
        r
    }

    fn emit(&mut self, op: u32, dst: u32, a1: u32, a2: u32) {
        self.code.extend_from_slice(&[op, dst, a1, a2]);
    }

    pub fn load_col_int(&mut self, col_idx: usize) -> u32 {
        let dst = self.alloc_reg();
        self.emit(EXPR_LOAD_COL_INT, dst, col_idx as u32, 0);
        dst
    }

    pub fn load_col_float(&mut self, col_idx: usize) -> u32 {
        let dst = self.alloc_reg();
        self.emit(EXPR_LOAD_COL_FLOAT, dst, col_idx as u32, 0);
        dst
    }

    /// `value as u32` — lower 32 bits; sufficient for typical filter constants.
    pub fn load_const(&mut self, value: i64) -> u32 {
        let dst = self.alloc_reg();
        self.emit(EXPR_LOAD_CONST, dst, value as u32, 0);
        dst
    }

    pub fn add(&mut self, a: u32, b: u32) -> u32 {
        let dst = self.alloc_reg();
        self.emit(EXPR_INT_ADD, dst, a, b);
        dst
    }

    pub fn sub(&mut self, a: u32, b: u32) -> u32 {
        let dst = self.alloc_reg();
        self.emit(EXPR_INT_SUB, dst, a, b);
        dst
    }

    pub fn cmp_eq(&mut self, a: u32, b: u32) -> u32 {
        let dst = self.alloc_reg();
        self.emit(EXPR_CMP_EQ, dst, a, b);
        dst
    }

    pub fn cmp_ne(&mut self, a: u32, b: u32) -> u32 {
        let dst = self.alloc_reg();
        self.emit(EXPR_CMP_NE, dst, a, b);
        dst
    }

    pub fn cmp_gt(&mut self, a: u32, b: u32) -> u32 {
        let dst = self.alloc_reg();
        self.emit(EXPR_CMP_GT, dst, a, b);
        dst
    }

    pub fn cmp_ge(&mut self, a: u32, b: u32) -> u32 {
        let dst = self.alloc_reg();
        self.emit(EXPR_CMP_GE, dst, a, b);
        dst
    }

    pub fn cmp_lt(&mut self, a: u32, b: u32) -> u32 {
        let dst = self.alloc_reg();
        self.emit(EXPR_CMP_LT, dst, a, b);
        dst
    }

    pub fn cmp_le(&mut self, a: u32, b: u32) -> u32 {
        let dst = self.alloc_reg();
        self.emit(EXPR_CMP_LE, dst, a, b);
        dst
    }

    pub fn bool_and(&mut self, a: u32, b: u32) -> u32 {
        let dst = self.alloc_reg();
        self.emit(EXPR_BOOL_AND, dst, a, b);
        dst
    }

    pub fn bool_or(&mut self, a: u32, b: u32) -> u32 {
        let dst = self.alloc_reg();
        self.emit(EXPR_BOOL_OR, dst, a, b);
        dst
    }

    pub fn bool_not(&mut self, a: u32) -> u32 {
        let dst = self.alloc_reg();
        self.emit(EXPR_BOOL_NOT, dst, a, 0);
        dst
    }

    pub fn is_null(&mut self, col_idx: usize) -> u32 {
        let dst = self.alloc_reg();
        self.emit(EXPR_IS_NULL, dst, col_idx as u32, 0);
        dst
    }

    pub fn is_not_null(&mut self, col_idx: usize) -> u32 {
        let dst = self.alloc_reg();
        self.emit(EXPR_IS_NOT_NULL, dst, col_idx as u32, 0);
        dst
    }

    pub fn build(self, result_reg: u32) -> ExprProgram {
        ExprProgram {
            num_regs:   self.next_reg,
            result_reg,
            code:       self.code,
        }
    }
}
